compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -285,18 +285,15 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
                             // The rhs of this `normalizes-to` must always be an unconstrained infer var as it is
                             // the hack used by `normalizes-to` to ensure that every `normalizes-to` behaves the same
                             // regardless of the rhs.
-                            //
-                            // However it is important not to unconditionally replace the rhs with a new infer var
-                            // as otherwise we may replace the original unconstrained infer var with a new infer var
-                            // and never propagate any constraints on the new var back to the original var.
-                            let term = this
-                                .term_is_fully_unconstrained(goal)
-                                .then_some(goal.predicate.term)
-                                .unwrap_or_else(|| {
-                                    this.next_term_infer_of_kind(goal.predicate.term)
-                                });
+
+                            // NOTE: This term should never be considered unconstrained, not least due to the fact
+                            // that the universe index has been bumped up at least once by `evaluate_goal` above.
```

Comment:
why does `evaluate_goal` bump the universe index/alternatively is `term_is_fully_unconstrained` too pessimistic?

this feels worrying to me

Reply:
> I guess maybe we should change that code to replace the rhs with a new infer var, call evaluate_goal for the normalizes_to hack and then emit add an eq goal for the inference vars or sth? that feels safer to me

yeah i can do that, i have a branch that does that and coincidentally the code actually gets a bit cleaner too...